### PR TITLE
SER-841 Prevent click word from hanging

### PR DESCRIPTION
**Pull request by sam-dixon**
Catches error if "click <word>" command produces an invalid XPath expression.

### CODE REVIEW

**Review by MattWiethoff — CHANGES_REQUESTED** [review]
Spoke offline about this. Looks like we need to be escaping the search string / not passing in symbols.

**Comment by sam-dixon** [comment]
It turns out there were two separate issues with characters that needed to be escaped:
* the regex wouldn't match any phrases that contained regex special characters (this is why "click dot" failed)
* the `document.evaluate` call would hang for contractions, since adding any path with the "'" character would make the XPath expression invalid

Both of these should be fixed

File: src/content/overlay.ts [hunk 1]
```diff
@@ -37,14 +37,20 @@ export default class Overlay {
...
         document,
         null,
         XPathResult.ORDERED_NODE_SNAPSHOT_TYPE,
         null
       );
 
-      const re = new RegExp("\\b" + path.split(" ").join("\\s*\\b") + "\\b", "i");
+      const re = new RegExp(
+        path
+          .replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
```
Comment (MattWiethoff) [inline src/content/overlay.ts:52]:
```suggestion
          // See https://stackoverflow.com/questions/3561493/is-there-a-regexp-escape-function-in-javascript/3561711#3561711
          .replace(/[.*+?^${}()|[\]\\]/g, "\\$&")
```

File: src/content/overlay.ts [hunk 2]
```diff
@@ -37,14 +37,20 @@ export default class Overlay {
       // Look for elements with matching containing text, input elements with matching placeholder text, or img elements
       // with matching alt text.
       const snapshot = document.evaluate(
-        `.//*[not(self::script)][not(self::noscript)][not(self::title)][not(self::meta)][not(self::svg)][not(self::style)][text()[contains(translate(., 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), '${path}')]]|//input[contains(translate(@placeholder, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), '${path}')]|//img[contains(translate(@alt, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), '${path}')]`,
+        `.//*[not(self::script)][not(self::noscript)][not(self::title)][not(self::meta)][not(self::svg)][not(self::style)][text()[contains(translate(., 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), "${path}")]]|//input[contains(translate(@placeholder, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), "${path}")]|//img[contains(translate(@alt, 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz'), "${path}")]`,
```
Comment (MattWiethoff) [inline src/content/overlay.ts]: Nice find with both of these. While still rare, it looks like `"` can still cause issues.

@tmacwill do you have any view on this? Looks like it's hard to escape xpath. Maybe there's a better pattern here? Perhaps not trying to do all of this complicated logic in `document.evaluate`?

**Review by MattWiethoff — CHANGES_REQUESTED** [review]

**Comment by sam-dixon** [comment]
Just added a utility function to escape the quotes when building the XPath expression.

**Review by MattWiethoff — APPROVED** [review]
Great, thanks!

**Pull request merged**